task/10-katas-1-tasks.js

```diff
@@ -16,9 +16,90 @@
  *     { abbreviation : 'NbW',   azimuth : 348.75 }
  *  ]
  */
+
 function createCompassPoints() {
-    throw new Error('Not implemented');
     var sides = ['N','E','S','W'];  // use array of cardinal directions only!
+    class CompasPoint{
+        constructor(abbreviation, azimuth){
+            this.azimuth = azimuth<0?360+azimuth:azimuth;
+            this.abbreviation = abbreviation;
+        }
+    }
+
+    class StartBuilder{
+        buildNorthAndSouth(firstChar){
+            return [new CompasPoint(firstChar, sides.indexOf(firstChar)*90)]
+            .concat(new MiddleBuilder().buildNorthAndSouth(firstChar));
+        }
+        buildEastAndWest(firstChar){
+            return [new CompasPoint(firstChar, sides.indexOf(firstChar)*90)]
+            .concat(new MiddleBuilder().buildEastAndWest(firstChar));
+        }
+        
+    }
+
+    class MiddleBuilder{
+        buildNorthAndSouth(firstChar){
+            return new EndBuilder().build(firstChar+'b', 'b')
+            .concat(new EndBuilder().build(firstChar+firstChar, firstChar))
+            .concat(new EndBuilder().build(firstChar+sides[(sides.indexOf(firstChar)+1)%4], firstChar))
+            .concat(new EndBuilder().build(firstChar+sides[(sides.indexOf(firstChar)+3)%4], firstChar));
+        }
+        buildEastAndWest(firstChar){
+            let prevDirection = sides[(sides.indexOf(firstChar)+3)%4];
+            let nextDirection = sides[(sides.indexOf(firstChar)+1)%4];
+            return new EndBuilder().build(firstChar+'b', 'b')
+            .concat([new CompasPoint(firstChar+nextDirection+firstChar, 
+                                     sides.indexOf(firstChar)*90+22.5*this.countSign(firstChar, nextDirection))])
+            .concat([new CompasPoint(firstChar+prevDirection+firstChar, 
+                                     sides.indexOf(firstChar)*90+22.5*this.countSign(firstChar, prevDirection))])
+        }
+        
+        countSign(firstChar, curChar){
+            return Math.sign((sides.indexOf(curChar)+sides.indexOf(firstChar))%4-(sides.indexOf(curChar)+sides.indexOf(firstChar)+2)%4);
+        }
+        
+    }
+
+    class EndBuilder{
+        build(begin, endChar){
+            let startCharIndex = sides.indexOf(begin.charAt(0))
+            let prevDirection = sides[(sides.indexOf(endChar)+3)%4];
+            let nextDirection = sides[(sides.indexOf(endChar)+1)%4];
+            let prevDirectionForb = sides[(startCharIndex+3)%4];
+            let nextDirectionForb = sides[(startCharIndex+1)%4];
+            return begin.charAt(0) == begin.charAt(1) ? [new CompasPoint(begin+nextDirection, 
```

Comment:
"operator === is always preferable
sometimes ''==' may cause an error
(check at console :
'0' == 0
and '0' === 0)"